modules/apps/29-fee/keeper/escrow.go

```diff
@@ -96,6 +96,10 @@ func (k Keeper) DistributePacketFeesOnTimeout(ctx sdk.Context, timeoutRelayer sd
 // If the distribution fails for any reason (such as the receiving address being blocked),
 // the state changes will be discarded.
 func (k Keeper) distributeFee(ctx sdk.Context, receiver sdk.AccAddress, fee sdk.Coins) {
+	if k.IsLocked(ctx) {
```

Comment:
Is this line even necessary? I believe in both Ack and Timeout we're checking locked before we even call this function.

Why don't we just add it as CONTRACT that we must check `locked` before calling this function?

Reply:
When we distribute fees, we do it one by one. It is possible distributing one of the fees causes the escrow balance to go negative at which point we want to stop any distribution. 

For example, the escrow balance has 10 coins. 
recv fee - 12 coins
ack fee - 5 coins
timeout fee - 5 coins

We start with recv fee, see the balance would be negative - we want to stop distribution. If I don't have this call, then ack fee and timeout fee would be paid out, unless I brought the check up to both `DistributePacketFees` and `DistributePacketFeesOnTimeout` - each distribution event would need this check and would need to return a nil error which can be confusing from a code point of view

Reply:
I also generally don't like the idea of function contracts. They are unenforceable and more prone to misuse

Reply:
Yes it is. The module may become locked during fee distribution. I'd have to add checks  around every call to distribute fee. If the balance becomes negative due to locking, I'd need to return immediately

I find this to be much easier to reason about from a code point of view. I also don't like the idea of function contracts since they are unenforceable and prone to misuse

Reply:
I think I should actually go with a solution similar to this https://github.com/cosmos/ibc-go/pull/1029/commits/2b9aa5d69d2ead8c3027d7a05df9a5982143bb19

Since we now have multiple fees per ID, we probably want to revert all state changes if a negative balance would be reached and then return immediately

Reply:
Thoughts?

Reply:
Agreed, I think caching and discarding a partially executed tx is much cleaner.

We should cache the context and if the balance goes negative. Let's discard state changes and commit just the write to the locked flag